token-wrap/program/src/processor.rs

```diff
@@ -1,26 +1,915 @@
+            &spl_token_2022::ID,
+            wrapped_mint_account.key,
+            &get_wrapped_mint_authority(wrapped_mint_account.key),
+            freeze_authority.as_ref(),
+            unwrapped_mint_unpacked.decimals,
+        )?;
+        invoke_signed(
+            &init_mint_ix,
+            &[
+                wrapped_mint_account.clone(),
+                token_program.clone(),
+            ],
+            &[wrapped_mint_authority_seeds],
+        )?;
+        },
+        _ => return Err(ProgramError::InvalidAccountData),
+    }
+
+
+    // Create and Initialize Backpointer Account
+    let backpointer_lamports = rent.minimum_balance(std::mem::size_of::<Backpointer>());
+    let (_, bump) = get_wrapped_mint_backpointer_address_with_seed(unwrapped_mint_account.key);
+    let bump = [bump];
+    let signer_seeds: &[&[u8]] = &get_wrapped_mint_backpointer_address_signer_seeds(unwrapped_mint_account.key, &bump);
+   
+    let create_backpointer_account_ix = system_instruction::create_account(
+        funding_account.key,
+        backpointer_account.key,
+        backpointer_lamports,
+        std::mem::size_of::<Backpointer>() as u64,
+        program_id,
+    );
+    msg!("create_backpointer_account_ix {:?}", create_backpointer_account_ix);
+    invoke_signed(
+        &create_backpointer_account_ix,
+        &[
+            funding_account.clone(),
+            backpointer_account.clone(),
+            system_program.clone(),
+        ],
+        &[signer_seeds],
+    )?;
+
+    // Initialize Backpointer Account
+    let backpointer_data = Backpointer {
+        unwrapped_mint: *unwrapped_mint_account.key,
+    };
+    let data = &mut backpointer_account.data.borrow_mut();
+    let backpointer_data_bytes = bytemuck::bytes_of(&backpointer_data);
+    
+    for (i, byte) in backpointer_data_bytes.iter().enumerate() {
+        data[i] = *byte;
+    }
+    Ok(())
+}
+
+/// Process 'Wrap' instruction
+fn process_wrap(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let user_source_account = next_account_info(account_info_iter)?;
+    let escrow_account = next_account_info(account_info_iter)?;
+    let unwrapped_mint = next_account_info(account_info_iter)?;
+    let wrapped_mint = next_account_info(account_info_iter)?;
+    let user_destination_account = next_account_info(account_info_iter)?;
+    let escrow_mint_authority = next_account_info(account_info_iter)?;
+    let unwrapped_token_program = next_account_info(account_info_iter)?;
+    let wrapped_token_program = next_account_info(account_info_iter)?;
+    msg!("unwrapped_token_program {:?}", unwrapped_token_program.key);
+    msg!("wrapped_token_program {:?}", wrapped_token_program.key);
+
+    let signer = next_account_info(account_info_iter)?;
+    let escrow_unwrapped_account = spl_token_2022::state::Account::unpack(&escrow_account.data.borrow())?;
```

Comment:
We should be able to avoid this deserialization, and let the token programs take care of the checks